### DISCUSSION TITLE
'ERR_MODULE_NOT_FOUND' on Vercel Serverless API typescript

### DISCUSSION THREAD

**Discussion by simone98dm** [vercel/community#1226] [Help] [answered]
### Summary

I've created a Vite project (Vue3 and Typescript) and I've also created different endpoints that allow retrieving some data.

My idea is to have a `\api` folder with all the endpoints, and another `\backend` with all the business logic, so the file system would be like this:
```
api\
    index.ts
backend\
    enum.ts
    consts.ts
    logic.ts
src\
    ... vue files ...

.eslintrc.json
.prettierrc
tsconfig.json
...
```

But, recently I've deployed a new version of the app and the serverless API (`api\index.ts`) stop working because it cannot find the module `../backend/const.ts`.

The stack trace is the following:
```
2023-01-07T09:43:18.189Z	XXX	ERROR	Error: Cannot find module '/var/task/backend/const' imported from /var/task/api/index.js
    at new NodeError (node:internal/errors:393:5)
    at finalizeResolution (node:internal/modules/esm/resolve:328:11)
    at moduleResolve (node:internal/modules/esm/resolve:965:10)
    at moduleResolveWithNodePath (node:internal/modules/esm/resolve:909:12)
    at defaultResolve (node:internal/modules/esm/resolve:1173:79)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:841:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:40)
    at link (node:internal/modules/esm/module_job:75:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

In the `index.ts`, to import all the needed constants and functions, I'm using the notation `import { a } from '../backend/const'`, but it appears that is not valid anymore. The references are correct, also because they worked since yesterday after several deployments and they are working in my local environment but seems to be broken when I deploy the project in Vercel.

Also, after some try the problem just changes from one file to another, so even if the `const` file seems to be fixed, the same problem replicates to another file `enum`, and so on...

- I've already tried to move all the files inside the `api` folder (in a single file to not exceed the maximum api endpoint allowed) but it didn't work.
- I've tried to update the references, but it didn't work.
- I've tried to re-deploy everything, but it didn't work.
- I've tried to change the node version (from `16` -> `18`), but it didn't work.
- I've tried to change the `target` in `tsconfig.json` from `ESNext` to `ES2019`, but it didn't work.
- Remove the `"type": "module"` from `package.json` , but it didn't work, also if I remove this property the Vue project get broken

Does somebody have some ideas to fix or debug this problem?

### Additional information

My `tsconfig.json`

```json
{
  "compilerOptions": {
    "target": "ESNext",
    "useDefineForClassFields": true,
    "module": "ESNext",
    "moduleResolution": "Node",
    "strict": true,
    "jsx": "preserve",
    "sourceMap": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "esModuleInterop": true,
    "lib": [
      "ESNext",
      "DOM"
    ],
    "baseUrl": ".",
    "skipLibCheck": true,
    "paths": {
      "~/*": [
        "src/*"
      ],
      "*": [
        "types/*.d.ts"
      ]
    },
    "types": []
  },
  "include": [
    "src/**/*.ts",
    "src/**/*.d.ts",
    "src/**/*.tsx",
    "src/**/*.vue",
    "meta.ts"
  ],
  "references": [
    {
      "path": "./tsconfig.node.json"
    }
  ],
  "exclude": [
    "src/main/test/cypress",
  ]
}
```

**Comment by simone98dm** [comment]
Update: I've run the command `vercel build` to check what would be the output and it seems that the functions in `.vercel/output` folder have the correct references to the files. So, in the end, I don't really know where the problem is😔

**Comment by tmilar** [comment]
Looks like something broke/changed in Vercel-side build process? I started getting a similar issue since friday, as being discussed in this other thread: https://github.com/vercel/community/discussions/1225.
In our case it's blocking us from generating any new deployment, even for projects that were already in production without doing any change in the code.

**Reply by amyegan** [reply]
Thanks for linking to the other discussion! 

@simone98dm Does the build work if you [add an environment variable](https://vercel.com/docs/concepts/projects/environment-variables) with the name `VERCEL_CLI_VERSION` and value `vercel@28.10.2`?

**Reply by simone98dm** [reply]
Yes, it seems to work now with version `28.10.2`, I'll check again if the error persists with the next version. Thank you✌️

**Reply by amyegan** [reply, marked as answer]
Update from the other thread: https://github.com/vercel/community/discussions/1225#discussioncomment-4642381

In it you will find a TS config change to try, a reminder to remove the CLI version environment variable to ensure you get future updates, and a request to share some specific information with [Vercel Support](https://vercel.com/help#issues) if you are still facing issues with the config change. 🙇

**Reply by tech-daedalus** [reply]
@amyegan the only problem left seem seems to be that Vercel Serverless APIs are not working with typescript paths

**Comment by bhaskarmurthy** [comment]
Hi - I'm running into the same issue, and have the same resolution - downgrading to `28.10.2` solves the serverless APi function error.
My `tsconfig.json` and `package.json` configs have the appropriate values - the build doesn't work with `28.10.3` and does work with `28.10.3` with the same config.

**Comment by abernh** [comment]
Downgrading the CLI to `28.10.2` did *not* solve my `ERR_MODULE_NOT_FOUND` issue (running NUXT in SSG with a small API).
Throwing `type: module` into my `package.json` also just caused some linter messages to go off, so I'm not happy with that either.

My "fix" to unblock our deployments for now was to pre-build my API endpoints into single-files and now I'm waiting for Vercel (support) to come up with a real fix for this situation.

**Reply by amyegan** [reply]
Hi @abernh! If downgrading the CLI to `28.10.2` didn't solve the issue for you, then I think you have a different problem. Can you [start a new discussion](https://github.com/vercel/community/discussions/new) with some more info about your app and the error thrown?

**Comment by scottbasgaard** [comment]
Downgrading to `28.10.2` fixed this for me as well. Thanks for helping in the meantime and hope this is fixed 🙏

**Comment by GustafHultgren** [comment]
Hi @amyegan ! I'm using Vue with Vite, and downgrading to `28.10.2` solved the problem for me back in February. Now I tried bumping my vercel cli version to 31.0.4, but it seems the same serverless import issue is still there. The docs are hinting at using the community plugin `vite-plugin-vercel`. Is this the only way, or will there be some official support coming?

**Reply by amyegan** [reply]
Hey @GustafHultgren. Basically every framework is supported via the [Build Output API](https://vercel.com/docs/build-output-api/v3). If build output is configured to match the spec, then it will work with Vercel's features.

Vite sites that use SSR or Serverless Functions require some complex configuration that many people prefer not to do on their own. You can configure it yourself if you prefer, but the community plugins can make it faster and easier to do.

**Comment by dbssman** [comment]
Hi @amyegan 

January 2024 and this issue still persists.

Importing a typescript module into a vercel serverless function doesn't work ootb without downgrading to @28.10.2

Is this expected ?

**Comment by qoomon** [comment]
Seems to work, if you change `import { a } from '../backend/const'` to `import { a } from '../backend/const.js'`
Although it seems odd it works as expected

**Reply by dbssman** [reply]
Not the case for typescript files

**Reply by qoomon** [reply]
yes it does

**Reply by dbssman** [reply]
Can you share your vercel CLI version ?

**Reply by qoomon** [reply]
I don't use the cli i just push to my repository see https://github.com/qoomon/github-actions-access-token/tree/main/server/deployments/vercel

**Reply by dbssman** [reply]
Can you try to import a ts file from your api ts file, because your ts file is just importing a js file

**Reply by qoomon** [reply]
yes it is actually a ts file see https://github.com/qoomon/github-actions-access-token/blob/main/server/app.ts

**Comment by CryptoGraffe** [comment]
I'm stuck on this as well. I don't want to change 100s of imports to .js . Anyone have any other tips?

**Reply by CryptoGraffe** [reply]
A crazy thing about this is that editors don't auto add the .js to imports, and as far as I can tell, there is no way to change this in a tsconfig or otherwise.

**Reply by qoomon** [reply]
@CryptoGraffe have look at ESLint plugin https://github.com/solana-labs/eslint-plugin-require-extensions
then just run `eslint . --fix`
